### PR TITLE
Replaces all instances of spawn() with addtimer/INVOKE_ASYNC

### LINKED ISSUE #27136: Spawn leaves reference to object it's called on, preventing garbage collection
More of a BYOND issue/quirk than codebase related, but still important enough I figured an issue report would be warranted.

Either case, yet another reason to hate/disallow use of `spawn`. Whenever `spawn` is used, it leaves an external reference to the object it's called on, which will prevent the object in question from garbage collecting. 

For instance:

```
/obj/widget/New()
	spawn(600)
		qdel(src)
```

If something were to delete `widget` before spawn ran, then `widget` would fail to soft delete/GC.

Moral of the story: don't use spawn as a timer function, as it'll prevent GCing under some circumstances and therefore increase total CPU cost of deletion.

### PR DESCRIPTION
**Pull request by silicons**
some things to note:
rnd stuff will be done in the rnd pr because i don't want to give myself too many conflicts.
i don't know if doing it to the tooltips is a great idea
the disposals spawns semm needless since everything is already made at initialize unlike before when it used new() so i guess there shouldn't be problems?
Fixes #27136 as after this only R&D consoles will be the problem and I left a comment there.

### CODE REVIEW

File: code/modules/reagents/reagent_containers/spray.dm [hunk 1]
```diff
@@ -73,40 +73,42 @@
...
-							D.reagents.reaction(M, VAPOR)
-							puff_reagent_left -= 1
-					else if(!range_left)
-						D.reagents.reaction(T, VAPOR)
-				else
-					D.reagents.reaction(T, VAPOR)
-					if(ismob(T))
+	INVOKE_ASYNC(src, .proc/do_spray, A, wait_step, D, range, puff_reagent_left)
+
+/obj/item/weapon/reagent_containers/spray/proc/do_spray(atom/A, wait_step, obj/effect/decal/chempuff/D, range, puff_reagent_left)
```
Comment (Cyberboss) [inline code/modules/reagents/reagent_containers/spray.dm]: Use set waitfor = FALSE

File: code/modules/recycling/disposal-unit.dm [hunk 2]
```diff
@@ -379,6 +379,10 @@
 	else if(full_pressure)
 		add_overlay("dispover-ready")
 
+/obj/machinery/disposal/bin/proc/do_flush()
+	SSblackbox.inc("disposal_auto_flush",1)
```
Comment (Cyberboss) [inline code/modules/recycling/disposal-unit.dm]: no need for ,1

File: code/modules/recycling/disposal-unit.dm [hunk 3]
```diff
@@ -379,6 +379,10 @@
 	else if(full_pressure)
 		add_overlay("dispover-ready")
 
+/obj/machinery/disposal/bin/proc/do_flush()
```
Comment (Cyberboss) [inline code/modules/recycling/disposal-unit.dm]: Use set waitfor = FALSE

File: code/modules/recycling/disposal-unit.dm [hunk 4]
```diff
@@ -389,9 +393,7 @@
 	if(flush_count >= flush_every_ticks)
 		if(contents.len)
 			if(full_pressure)
-				spawn(0)
-					SSblackbox.inc("disposal_auto_flush",1)
-					flush()
+				INVOKE_ASYNC(src, .proc/do_flush)
```
Comment (Cyberboss) [inline code/modules/recycling/disposal-unit.dm]: standard call

File: code/modules/reagents/reagent_containers/spray.dm [hunk 5]
```diff
@@ -73,40 +73,42 @@
...
-						var/mob/M = T
-						if(!M.lying || !range_left)
-							D.reagents.reaction(M, VAPOR)
-							puff_reagent_left -= 1
-					else if(!range_left)
-						D.reagents.reaction(T, VAPOR)
-				else
-					D.reagents.reaction(T, VAPOR)
-					if(ismob(T))
+	INVOKE_ASYNC(src, .proc/do_spray, A, wait_step, D, range, puff_reagent_left)
```
Comment (Cyberboss) [inline code/modules/reagents/reagent_containers/spray.dm]: standard call

File: code/modules/research/destructive_analyzer.dm [hunk 6]
```diff
@@ -59,11 +59,13 @@ Note: Must be placed within 3 tiles of the R&D Console
...
 		O.forceMove(src)
 		to_chat(user, "<span class='notice'>You add the [O.name] to the [src.name]!</span>")
 		flick("d_analyzer_la", src)
-		spawn(10)
-			icon_state = "d_analyzer_l"
-			busy = 0
+		addtimer(CALLBACK(src, .proc/finish_loading), 10)
+
+/obj/machinery/r_n_d/destructive_analyzer/proc/finish_loading()
+	icon_state = "d_analyzer_l"
```
Comment (Cyberboss) [inline code/modules/research/destructive_analyzer.dm]: use update_icon

File: code/modules/research/destructive_analyzer.dm [hunk 7]
```diff
@@ -59,11 +59,13 @@ Note: Must be placed within 3 tiles of the R&D Console
...
+	busy = 0
```
Comment (Cyberboss) [inline code/modules/research/destructive_analyzer.dm]: FALSE

File: code/modules/research/experimentor.dm [hunk 8]
```diff
@@ -504,9 +506,11 @@
 			use_power(500000)
 			investigate_log("Experimentor has drained power from its APC", INVESTIGATE_EXPERIMENTOR)
 
-	spawn(resetTime)
-		icon_state = "h_lathe"
-		recentlyExperimented = 0
+	addtimer(CALLBACK(src, .proc/reset_exp), resetTime)
+
+/obj/machinery/r_n_d/experimentor/proc/reset_exp()
+	icon_state = "h_lathe"
```
Comment (Cyberboss) [inline code/modules/research/experimentor.dm]: update_icon

File: code/modules/research/experimentor.dm [hunk 9]
```diff
@@ -673,31 +679,37 @@
 		dupes |= R
 		R.throw_at(pick(oview(7,get_turf(src))),10,1)
 	counter = 0
-	spawn(rand(10,100))
-		for(counter = 1; counter <= dupes.len; counter++)
-			var/obj/item/weapon/relic/R = dupes[counter]
-			qdel(R)
+	addtimer(CALLBACK(src, .proc/del_dupes, dupes, counter), rand(10, 100))
```
Comment (Cyberboss) [inline code/modules/research/experimentor.dm]: You can use that qdel_list_proc you made instead

File: code/modules/research/experimentor.dm [hunk 10]
```diff
@@ -673,31 +679,37 @@
...
-			visible_message("<span class='notice'>The [src]'s top opens, releasing a powerful blast!</span>")
-			explosion(user.loc, -1, rand(1,5), rand(1,5), rand(1,5), rand(1,5), flame_range = 2)
-			warn_admins(user, "Explosion")
-			qdel(src) //Comment this line to produce a light grenade (the bomb that keeps on exploding when used)!!
+	addtimer(CALLBACK(src, .proc/do_explode, user), rand(35, 100))
+
+/obj/item/weapon/relic/proc/do_explode(mob/user)
+	if(loc == user)
+		visible_message("<span class='notice'>The [src]'s top opens, releasing a powerful blast!</span>")
+		explosion(user.loc, -1, rand(1,5), rand(1,5), rand(1,5), rand(1,5), flame_range = 2)
```
Comment (Cyberboss) [inline code/modules/research/experimentor.dm]: explosion(user, 0, ...

File: code/modules/research/experimentor.dm [hunk 11]
```diff
@@ -673,31 +679,37 @@
...
-		if(src.loc == user)
-			visible_message("<span class='notice'>The [src]'s top opens, releasing a powerful blast!</span>")
-			explosion(user.loc, -1, rand(1,5), rand(1,5), rand(1,5), rand(1,5), flame_range = 2)
-			warn_admins(user, "Explosion")
-			qdel(src) //Comment this line to produce a light grenade (the bomb that keeps on exploding when used)!!
+	addtimer(CALLBACK(src, .proc/do_explode, user), rand(35, 100))
+
+/obj/item/weapon/relic/proc/do_explode(mob/user)
+	if(loc == user)
+		visible_message("<span class='notice'>The [src]'s top opens, releasing a powerful blast!</span>")
```
Comment (Cyberboss) [inline code/modules/research/experimentor.dm]: \The

File: code/modules/research/experimentor.dm [hunk 12]
```diff
@@ -673,31 +679,37 @@
...
-			visible_message("<span class='notice'>The [src] twists and bends, relocating itself!</span>")
-			throwSmoke(userturf)
-			do_teleport(user, userturf, 8, asoundin = 'sound/effects/phasein.ogg')
-			throwSmoke(get_turf(user))
-			warn_admins(user, "Teleport", 0)
+	addtimer(CALLBACK(src, .proc/do_teleport, user), rand(10, 30))
+
+/obj/item/weapon/relic/proc/do_teleport(mob/user)
+	var/turf/userturf = get_turf(user)
+	if(src.loc == user && userturf.z != ZLEVEL_CENTCOM) //Because Nuke Ops bringing this back on their shuttle, then looting the ERT area is 2fun4you!
```
Comment (Cyberboss) [inline code/modules/research/experimentor.dm]: remove src

File: code/modules/research/rdconsole.dm [hunk 13]
```diff
@@ -29,6 +29,8 @@ won't update every console in existence) but it's more of a hassle to do. Also,
 
 */
 
+//SPAWN() removal pending in RND techweb PR.
+
```
Comment (Cyberboss) [inline code/modules/research/rdconsole.dm]: Comment unecessary

File: code/modules/spells/spell_types/dumbfire.dm [hunk 14]
```diff
@@ -33,58 +36,62 @@
...
-				M.Turn(dir2angle(projectile.dir))
-				projectile.transform = M
-
-				sleep(proj_step_delay)
-
-			if(projectile)
-				qdel(projectile)
+		INVOKE_ASYNC(src, .proc/launch_at, target, user)
+
+/obj/effect/proc_holder/spell/dumbfire/proc/launch_at(turf/target, mob/user)
```
Comment (Cyberboss) [inline code/modules/spells/spell_types/dumbfire.dm]: Use set waitfor = FALSE

File: code/modules/spells/spell_types/dumbfire.dm [hunk 15]
```diff
@@ -33,58 +36,62 @@
...
-				current_loc = projectile.loc
-				var/matrix/M = new
-				M.Turn(dir2angle(projectile.dir))
-				projectile.transform = M
-
-				sleep(proj_step_delay)
-
-			if(projectile)
-				qdel(projectile)
+		INVOKE_ASYNC(src, .proc/launch_at, target, user)
```
Comment (Cyberboss) [inline code/modules/spells/spell_types/dumbfire.dm]: standard call

File: code/modules/spells/spell_types/dumbfire.dm [hunk 16]
```diff
@@ -33,58 +36,62 @@
...
+
+/obj/effect/proc_holder/spell/dumbfire/proc/launch_at(turf/target, mob/user)
+	var/obj/effect/proc_holder/spell/targeted/projectile
+	if(istext(proj_type))
+		var/projectile_type = text2path(proj_type)
+		projectile = new projectile_type(user)
+	else if(istype(proj_type,/obj/effect/proc_holder/spell))
+		projectile = new /obj/effect/proc_holder/spell/targeted/trigger(user)
+		projectile:linked_spells += proj_type
```
Comment (Cyberboss) [inline code/modules/spells/spell_types/dumbfire.dm]: please remove this colon if you can

File: code/modules/spells/spell_types/dumbfire.dm [hunk 17]
```diff
@@ -33,58 +36,62 @@
...
+			projectile.cast(current_loc,user=user)
+			break
+
+		var/mob/living/L = locate(/mob/living) in range(projectile, proj_trigger_range) - user
+		if(L && L.stat != DEAD)
+			projectile.cast(L.loc,user=user)
+			break
+
+		if(proj_trail && projectile)
+			INVOKE_ASYNC(src, .proc/proj_trail, projectile)
```
Comment (Cyberboss) [inline code/modules/spells/spell_types/dumbfire.dm]: standard call

File: code/modules/spells/spell_types/dumbfire.dm [hunk 18]
```diff
@@ -33,58 +36,62 @@
...
+		var/matrix/M = new
+		M.Turn(dir2angle(projectile.dir))
+		projectile.transform = M
+
+		sleep(proj_step_delay)
+
+	if(projectile)
+		qdel(projectile)
+
+/obj/effect/proc_holder/spell/dumbfire/proc/proj_trail(obj/effect/proc_holder/spell/targeted/projectile)
```
Comment (Cyberboss) [inline code/modules/spells/spell_types/dumbfire.dm]: Use set waitfor = FALSE

File: code/modules/spells/spell_types/dumbfire.dm [hunk 19]
```diff
@@ -33,58 +36,62 @@
...
+	if(projectile)
+		var/obj/effect/overlay/trail = new /obj/effect/overlay(projectile.loc)
+		trail.icon = proj_trail_icon
+		trail.icon_state = proj_trail_icon_state
+		trail.density = 0
```
Comment (Cyberboss) [inline code/modules/spells/spell_types/dumbfire.dm]: FALSE

File: code/modules/spells/spell_types/genetic.dm [hunk 20]
```diff
@@ -23,8 +23,10 @@
...
 		target.disabilities |= disabilities
-		spawn(duration)
-			if(target && !QDELETED(target))
-				for(var/A in mutations)
-					target.dna.remove_mutation(A)
-				target.disabilities &= ~disabilities
+		addtimer(CALLBACK(src, .proc/remove, target), duration)
+
+/obj/effect/proc_holder/spell/targeted/genetic/proc/remove(mob/living/carbon/target)
+	if(target && !QDELETED(target))
```
Comment (Cyberboss) [inline code/modules/spells/spell_types/genetic.dm]: just !QDELETED(target)

File: code/modules/spells/spell_types/projectile.dm [hunk 21]
```diff
@@ -25,62 +28,66 @@
...
-					projectile.perform(list(target),user=user)
-					break
-
-				current_loc = projectile.loc
-
-				sleep(proj_step_delay)
-
-			if(projectile)
-				qdel(projectile)
+		INVOKE_ASYNC(src, .proc/launch, target, user)
```
Comment (Cyberboss) [inline code/modules/spells/spell_types/projectile.dm]: standard call

File: code/modules/spells/spell_types/projectile.dm [hunk 22]
```diff
@@ -25,62 +28,66 @@
...
+
+/obj/effect/proc_holder/spell/targeted/projectile/proc/launch(mob/living/target, mob/user)
+	var/obj/effect/proc_holder/spell/targeted/projectile
+	if(istext(proj_type))
+		var/projectile_type = text2path(proj_type)
+		projectile = new projectile_type(user)
+	if(istype(proj_type,/obj/effect/proc_holder/spell))
+		projectile = new /obj/effect/proc_holder/spell/targeted/trigger(user)
+		projectile:linked_spells += proj_type
```
Comment (Cyberboss) [inline code/modules/spells/spell_types/projectile.dm]: colon

File: code/modules/spells/spell_types/projectile.dm [hunk 23]
```diff
@@ -25,62 +28,66 @@
...
+
+		if(!projectile) // step and step_to sleeps so we'll have to check again.
+			break
+
+		if(!target || (!proj_lingering && projectile.loc == current_loc)) //if it didn't move since last time
+			qdel(projectile)
+			break
+
+		if(proj_trail && projectile)
+			INVOKE_ASYNC(src, .proc/spawntrail)
```
Comment (Cyberboss) [inline code/modules/spells/spell_types/projectile.dm]: standard call

File: code/modules/spells/spell_types/projectile.dm [hunk 24]
```diff
@@ -25,62 +28,66 @@
...
+
+	if(projectile)
+		qdel(projectile)
+
+/obj/effect/proc_holder/spell/targeted/projectile/proc/spawntrail(obj/effect/proc_holder/spell/targeted/projectile)
+	if(projectile)
+		var/obj/effect/overlay/trail = new /obj/effect/overlay(projectile.loc)
+		trail.icon = proj_trail_icon
+		trail.icon_state = proj_trail_icon_state
+		trail.density = 0
```
Comment (Cyberboss) [inline code/modules/spells/spell_types/projectile.dm]: FALSE

File: code/modules/spells/spell_types/projectile.dm [hunk 25]
```diff
@@ -25,62 +28,66 @@
...
-
-				current_loc = projectile.loc
-
-				sleep(proj_step_delay)
-
-			if(projectile)
-				qdel(projectile)
+		INVOKE_ASYNC(src, .proc/launch, target, user)
+
+/obj/effect/proc_holder/spell/targeted/projectile/proc/launch(mob/living/target, mob/user)
```
Comment (Cyberboss) [inline code/modules/spells/spell_types/projectile.dm]: Use set waitfor = FALSE

File: code/modules/tooltip/tooltip.dm [hunk 26]
```diff
@@ -81,15 +81,16 @@ Notes:
...
 	if (src.queueHide)
-		spawn(1)
-			winshow(src.owner, src.control, 0)
+		addtimer(src, do_hide(), 1)
 	else
-		winshow(src.owner, src.control, 0)
+		do_hide()
 
-	src.queueHide = src.showing ? 1 : 0
+	queueHide = showing ? 1 : 0
```
Comment (Cyberboss) [inline code/modules/tooltip/tooltip.dm]: TRUE : FALSE

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/modules/spells/spell_types/dumbfire.dm (hunk 16 above)
Comment (kevinz000) [inline code/modules/spells/spell_types/dumbfire.dm]: can i go "i'll do this in a later pr"? this actually looks like a pain in the ass and would probably require a full refactor if i was to do it.

File: code/modules/spells/spell_types/projectile.dm (hunk 22 above)
Comment (kevinz000) [inline code/modules/spells/spell_types/projectile.dm]: refactor :^)

File: code/modules/spells/spell_types/dumbfire.dm (hunk 16 above)
Comment (Cyberboss) [inline code/modules/spells/spell_types/dumbfire.dm]: no

**Comment by Tacolizard** [comment]
There shouldn't be any reason to do this with tooltips.

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Comment by Cyberboss** [comment]
@Tacolizard if it uses spawn it has to go

File: code/modules/spells/spell_types/dumbfire.dm (hunk 18 above)
Comment (kevinz000) [inline code/modules/spells/spell_types/dumbfire.dm]: done

File: code/modules/recycling/disposal-unit.dm (hunk 3 above)
Comment (kevinz000) [inline code/modules/recycling/disposal-unit.dm]: done

File: code/modules/spells/spell_types/dumbfire.dm (hunk 18 above)
Comment (kevinz000) [inline code/modules/spells/spell_types/dumbfire.dm]: unless you mean remove the proc and just reinsert the code into the calling proc

File: code/modules/recycling/disposal-unit.dm (hunk 3 above)
Comment (kevinz000) [inline code/modules/recycling/disposal-unit.dm]: unless you mean remove the proc and just reinsert the code into the calling proc

File: code/modules/recycling/disposal-unit.dm [hunk 27]
```diff
@@ -389,15 +394,13 @@
...
-					SSblackbox.inc("disposal_auto_flush",1)
-					flush()
+				do_flush()
 		flush_count = 0
 
 	updateDialog()
 
 	if(flush && air_contents.return_pressure() >= SEND_PRESSURE) // flush can happen even without power
-		INVOKE_ASYNC(src, .proc/flush)
+		flush()
```
Comment (Cyberboss) [inline code/modules/recycling/disposal-unit.dm]: Why this?

File: code/modules/research/experimentor.dm [hunk 28]
```diff
@@ -299,7 +301,7 @@
 			investigate_log("Experimentor has released [chosenchem] smoke.", INVESTIGATE_EXPERIMENTOR)
 			var/datum/effect_system/smoke_spread/chem/smoke = new
 			smoke.set_up(R, 0, src, silent = 1)
-			playsound(src.loc, 'sound/effects/smoke.ogg', 50, 1, -3)
+			playsound(loc, 'sound/effects/smoke.ogg', 50, 1, -3)
```
Comment (Cyberboss) [inline code/modules/research/experimentor.dm]: playsound(src, ...

File: code/modules/research/experimentor.dm [hunk 29]
```diff
@@ -311,26 +313,26 @@
 			R.add_reagent(chosenchem , 50)
 			var/datum/effect_system/smoke_spread/chem/smoke = new
 			smoke.set_up(R, 0, src, silent = 1)
-			playsound(src.loc, 'sound/effects/smoke.ogg', 50, 1, -3)
+			playsound(loc, 'sound/effects/smoke.ogg', 50, 1, -3)
```
Comment (Cyberboss) [inline code/modules/research/experimentor.dm]: playsound(src,...

File: code/modules/research/experimentor.dm [hunk 30]
```diff
@@ -383,7 +385,7 @@
 		if(prob(EFFECT_PROB_LOW) && criticalReaction)
 			visible_message("<span class='warning'>[src]'s emergency coolant system gives off a small ding!</span>")
 			var/obj/item/weapon/reagent_containers/food/drinks/coffee/C = new /obj/item/weapon/reagent_containers/food/drinks/coffee(get_turf(pick(oview(1,src))))
-			playsound(src.loc, 'sound/machines/ding.ogg', 50, 1) //Ding! Your death coffee is ready!
+			playsound(loc, 'sound/machines/ding.ogg', 50, 1) //Ding! Your death coffee is ready!
```
Comment (Cyberboss) [inline code/modules/research/experimentor.dm]: playsound(src, ...

File: code/modules/research/experimentor.dm [hunk 31]
```diff
@@ -398,13 +400,13 @@
 			investigate_log("Experimentor has released frostoil gas.", INVESTIGATE_EXPERIMENTOR)
 			var/datum/effect_system/smoke_spread/chem/smoke = new
 			smoke.set_up(R, 0, src, silent = 1)
-			playsound(src.loc, 'sound/effects/smoke.ogg', 50, 1, -3)
+			playsound(loc, 'sound/effects/smoke.ogg', 50, 1, -3)
```
Comment (Cyberboss) [inline code/modules/research/experimentor.dm]: playsound(src, ...

File: code/modules/research/experimentor.dm [hunk 32]
```diff
@@ -433,14 +435,14 @@
 			new /obj/item/stack/sheet/plasteel(get_turf(pick(oview(1,src))))
 		else if(prob(EFFECT_PROB_VERYLOW-badThingCoeff))
 			visible_message("<span class='danger'>[src]'s crusher goes way too many levels too high, crushing right through space-time!</span>")
-			playsound(src.loc, 'sound/effects/supermatter.ogg', 50, 1, -3)
+			playsound(loc, 'sound/effects/supermatter.ogg', 50, 1, -3)
```
Comment (Cyberboss) [inline code/modules/research/experimentor.dm]: playsound(src, ...

File: code/modules/research/experimentor.dm [hunk 33]
```diff
@@ -433,14 +435,14 @@
...
-			playsound(src.loc, 'sound/effects/supermatter.ogg', 50, 1, -3)
+			playsound(loc, 'sound/effects/supermatter.ogg', 50, 1, -3)
 			investigate_log("Experimentor has triggered the 'throw things' reaction.", INVESTIGATE_EXPERIMENTOR)
 			for(var/atom/movable/AM in oview(7,src))
 				if(!AM.anchored)
 					AM.throw_at(src,10,1)
 		else if(prob(EFFECT_PROB_LOW-badThingCoeff))
 			visible_message("<span class='danger'>[src]'s crusher goes one level too high, crushing right into space-time!</span>")
-			playsound(src.loc, 'sound/effects/supermatter.ogg', 50, 1, -3)
+			playsound(loc, 'sound/effects/supermatter.ogg', 50, 1, -3)
```
Comment (Cyberboss) [inline code/modules/research/experimentor.dm]: playsound(src, ...

File: code/modules/research/experimentor.dm [hunk 34]
```diff
@@ -458,7 +460,7 @@
 
 	if(exp == SCANTYPE_DISCOVER)
 		visible_message("[src] scans the [exp_on], revealing its true nature!")
-		playsound(src.loc, 'sound/effects/supermatter.ogg', 50, 3, -1)
+		playsound(loc, 'sound/effects/supermatter.ogg', 50, 3, -1)
```
Comment (Cyberboss) [inline code/modules/research/experimentor.dm]: playsound(src, ...

File: code/modules/research/experimentor.dm [hunk 35]
```diff
@@ -473,13 +475,13 @@
 			ejectItem()
 		if(globalMalf > 16 && globalMalf < 35)
 			visible_message("<span class='warning'>[src] melts [exp_on], ian-izing the air around it!</span>")
-			throwSmoke(src.loc)
+			throwSmoke(loc)
 			if(trackedIan)
 				throwSmoke(trackedIan.loc)
-				trackedIan.loc = src.loc
+				trackedIan.loc = loc
```
Comment (Cyberboss) [inline code/modules/research/experimentor.dm]: forceMove

File: code/modules/research/experimentor.dm [hunk 36]
```diff
@@ -626,19 +635,19 @@
 	smoke.start()
 
 /obj/item/weapon/relic/proc/corgicannon(mob/user)
-	playsound(src.loc, "sparks", rand(25,50), 1)
+	playsound(loc, "sparks", rand(25,50), 1)
```
Comment (Cyberboss) [inline code/modules/research/experimentor.dm]: playsound(src, ...

File: code/modules/research/experimentor.dm [hunk 37]
```diff
@@ -626,19 +635,19 @@
...
 /obj/item/weapon/relic/proc/corgicannon(mob/user)
-	playsound(src.loc, "sparks", rand(25,50), 1)
+	playsound(loc, "sparks", rand(25,50), 1)
 	var/mob/living/simple_animal/pet/dog/corgi/C = new/mob/living/simple_animal/pet/dog/corgi(get_turf(user))
 	C.throw_at(pick(oview(10,user)), 10, rand(3,8), callback = CALLBACK(src, .throwSmoke, C))
 	warn_admins(user, "Corgi Cannon", 0)
 
 /obj/item/weapon/relic/proc/clean(mob/user)
-	playsound(src.loc, "sparks", rand(25,50), 1)
+	playsound(loc, "sparks", rand(25,50), 1)
```
Comment (Cyberboss) [inline code/modules/research/experimentor.dm]: playsound(src, ...

File: code/modules/research/experimentor.dm [hunk 38]
```diff
@@ -626,19 +635,19 @@
...
 	var/obj/item/weapon/grenade/chem_grenade/cleaner/CL = new/obj/item/weapon/grenade/chem_grenade/cleaner(get_turf(user))
 	CL.prime()
 	warn_admins(user, "Smoke", 0)
 
 /obj/item/weapon/relic/proc/flash(mob/user)
-	playsound(src.loc, "sparks", rand(25,50), 1)
+	playsound(loc, "sparks", rand(25,50), 1)
```
Comment (Cyberboss) [inline code/modules/research/experimentor.dm]: playsound(src, ...

File: code/modules/research/experimentor.dm [hunk 39]
```diff
@@ -673,31 +682,37 @@
...
 		R.throw_at(pick(oview(7,get_turf(src))),10,1)
 	counter = 0
-	spawn(rand(10,100))
-		for(counter = 1; counter <= dupes.len; counter++)
-			var/obj/item/weapon/relic/R = dupes[counter]
-			qdel(R)
+	QDEL_LIST_IN(dupes, rand(10, 100))
 	warn_admins(user, "Rapid duplicator", 0)
 
+/obj/item/weapon/relic/proc/del_dupes(list/dupes, counter)
```
Comment (Cyberboss) [inline code/modules/research/experimentor.dm]: Unused proc

File: code/modules/tooltip/tooltip.dm [hunk 40]
```diff
@@ -81,15 +81,16 @@ Notes:
 
 /datum/tooltip/proc/hide()
 	if (src.queueHide)
-		spawn(1)
-			winshow(src.owner, src.control, 0)
+		addtimer(CALLBACK(src, .proc/do_hide), 1)
```
Comment (Cyberboss) [inline code/modules/tooltip/tooltip.dm]: TRUE

Comment (Cyberboss) [inline code/modules/tooltip/tooltip.dm]: Why does this have a param

File: code/modules/tooltip/tooltip.dm [hunk 41]
```diff
@@ -81,15 +81,16 @@ Notes:
...
+		do_hide()
 
-	src.queueHide = src.showing ? 1 : 0
+	queueHide = showing ? TRUE : FALSE
 
-	return 1
+	return TRUE
 
+/datum/tooltip/proc/do_hide()
+	winshow(owner, control, 0)
```
Comment (Cyberboss) [inline code/modules/tooltip/tooltip.dm]: FALSE

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/modules/tooltip/tooltip.dm (hunk 40 above)
Comment (kevinz000) [inline code/modules/tooltip/tooltip.dm]: it's not a param it's a delay of 1 for whatever reason 
it's gooncode or somethign ask @Tacolizard but I didn't want to touch it.

Comment (kevinz000) [inline code/modules/tooltip/tooltip.dm]: not bool

Comment (Tacolizard) [inline code/modules/tooltip/tooltip.dm]: i can't remember if this was gooncode or just copied from other tooltip shit. if you don't want to test it, I'd advise not touching it.

Comment (Tacolizard) [inline code/modules/tooltip/tooltip.dm]: or wait what I didn't write this I didn't realize what you mean by tooltips this is wirewraith(?)'s tooltip code.

**Comment by Cyberboss** [comment]
@kevinz000 fix the loc change

**Comment by kevinz000** [comment]
Ok give me a bit

**Comment by Cyberboss** [comment]
@kevinz000 !!!!!!!!

**Comment by kevinz000** [comment]
I JUST GOT TO A COMPUTER OK 6 HOURS I PROMISE

**Comment by kevinz000** [comment]
@Cyberboss FIXED

**Review by Cyberboss — APPROVED** [review]

**Pull request merged**